### PR TITLE
Add AS4 Fetching data article 

### PR DESCRIPTION
**Pull request by rkoron007**

### CODE REVIEW

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:56]:
```suggestion
    return await this.dbConnection.Booking.findByPk(bookingId);
```

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:47]: We should use `contextValue` here.

It could be interesting to make use of the `token` in our example, but not totally necessary. It does seem extraneous here if we're not using it anywhere.

Comment (trevor-scheer) [reply docs/source/data/fetching-data.mdx:47]: This also doesn't have to be called `contextValue` at all, we could just say the `constructor` needs a `token` as one of its options and leave it to the `context` function to pass it in correctly.

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:118]:
```suggestion
Your resolvers can then access your data sources from the shared `contextValue` object and use them to fetch data:
```

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:174]: For some reason this is creating a new empty line below this one and highlighting it in the preview. Maybe has to do with the comment on line 175?

File: docs/source/data/resolvers.mdx
Comment (trevor-scheer) [inline docs/source/data/resolvers.mdx:380]: Should this become `contextValue` or are we broadly keeping things as `context` in the docs?

File: docs/source/config.json
Comment (trevor-scheer) [inline docs/source/config.json:21]: I think this is a more generic article that should be listed before the REST counterpart

**Review by trevor-scheer — CHANGES_REQUESTED** [review]

File: docs/source/config.json
Comment (rkoron007) [reply docs/source/config.json:21]: I like that! I'll switch them :)

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:380]: I'll follow up with another subsequent PR for changing this to `contextValue` later. (This is linked in a _ton_ of places and I don't want to flood this PR with little link changes 😂 )

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:174]: Wait —in your deploy preview? The deploy preview looks fine for me?
Screen Shot 2022-09-27 at 11 04 16 AM

Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:174]: Fixed this!

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx]: Resolved this in my subsequent commit :)

File: docs/source/data/fetching-data.mdx
Comment (glasser) [reply docs/source/data/fetching-data.mdx]: Taking this one step further, I would suggest that `token` be passed directly as an option rather than being nested in `contextValue`.

The data source needs a token to run. The fact that the token is also embedded inside a GraphQL execution context value is not really relevant.

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:121]: This is only the case if you either pass something specific to the request to the data source construction or if the data source has caching, of course... (Which is why making data source setup explicit is helpful.) Not sure if that's helpful to clarify or not. (To be honest I find the entire article could be summarized by the single sentence "fetch data using the same mechanisms you'd use to fetch data in any other server; if the code you use to do that depends on properties of the request like authentication, you can put stuff on the context value" but I do think people are probably still helped by this article.)

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:143]: Maybe something like:

```suggestion
Apollo Server 3 contained [a small abstract class named `DataSource`](/apollo-server/v3/data/data-sources) that you could subclass for each of your data sources. Data sources were initialized in a special `dataSources` function instead of the main `context` function. In Apollo Server 4, you can create your data sources in the same `context` function as the rest of your per-request setup, and there is no special `DataSource` superclass. We recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
```

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:147]:
```suggestion
The below data source implementations extend the legacy [`DataSource` abstract class](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-datasource/src/index.ts), from the deprecated `apollo-datasource` package. Subclasses of `DataSource` define the logic required to communicate with a particular store or API. They may still be useful in Apollo Server 4, though you will have to manually call the `initialize` method on them after creating the object.
```

alternatively we could have a table of "useful data source classes for Apollo Server 4" and another table of "Apollo Server 3 data source classes that we hope will be updated for AS4 soon".

Or another organization we could do is:

- have a section that lists (modern) data source classes.  link to, say, RESTDataSource for now

- have a *separate* section talking about legacy Apollo Server 3 data sources; this one can have the text about how AS3 used to have this extra thing, and you can probably still use AS3 data sources with AS4 if you call `.initialize` yourself (but ideally the projects will update to the simpler AS4 API), and list them

... oh ok that is sort of what you're doing already, but I'd just separate "here are new things" from "here's a list of old things and the extra thing you have to do in order to use them even though their docs probably don't tell you to"

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:121]: Rephrased this as:

> If your data source class uses **caching** or its constructor requires an argument, the `context` function should create a _new instance_ of each data source class for each operation. Otherwise, you might accidentally combine results from multiple operations if those operations share a single data source instance.

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:143]: I like the feeling behind this! I did a few minor tweaks for style:
```suggestion
Apollo Server 3 contained [an abstract class named `DataSource`](/apollo-server/v3/data/data-sources) that each of your data sources could subclass. You'd then initialize each of your `DataSource` subclasses using a special `dataSources` function, which would attach your data sources to your `context` behind the scenes. 

In Apollo Server 4, you can now create your data sources in the same `context` function as the rest of your per-request setup, avoiding the `DataSource` superclass entirely. We recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
```

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:147]: So I do have a Note _right_ about this paragraph saying that they'll need to do some more work to get old  these subclasses working:

> > ⚠️ **Note**: The community built each data source package below for use with Apollo Server 3.  [As shown below](#using-datasource-subclasses), you can still use these packages in Apollo Server 4 with a bit of extra setup.

So I don't feel like we need the extra sentence you suggested if that sounds good to you.

For organization: we won't have more data source classes in the future, right?  We expect folks to make their data source classes—so having a separate section about "modern" classes feels a bit weird if it'll only be `RESTDataSource`?  I do agree that I could do a better job of separating out these older implementations, though. I could change the header of this section to "Apollo Server 3 Open-source implementations" if that'd feel like a good way to mark it as "old stuff"?

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx]: Good point! Updated to:
```ts
export class BookingsDataSource {
  private dbConnection;
  private token;

  constructor(options: { token: string }) {
    this.dbConnection = this.initializeDBConnection();
    this.token = options.contextValue.token;
  }

  async initializeDBConnection() {
    // set up our database details, instantiate our connection,
    // and return that database connection
    return dbConnection;
  }

  async getBooking(bookingId) {
    const user = await getUser(this.token);
    if (user) {
      return await this.dbConnection.Booking.findByPk(bookingId);
    } else {
      // handle invalid user
    }
  }

  //... more methods for finding and creating bookings
}
```

File: docs/source/data/fetching-data.mdx
Comment (glasser) [reply docs/source/data/fetching-data.mdx:121]: well I guess the issue is it's not caching in general, but "in memory caching of values that shouldn't be shared across requests".  (This is why it's a bit challenging that the REST DataSource page doesn't actually explain its two-layer caching model... you could link to that here if we had it.)

File: docs/source/data/fetching-data.mdx
Comment (glasser) [reply docs/source/data/fetching-data.mdx:143]: Looks good

File: docs/source/data/fetching-data.mdx
Comment (glasser) [reply docs/source/data/fetching-data.mdx:147]: I mean this comes back to the general issue that this is a page documenting a concept that doesn't really exist any more... Like if we're going to have a table of AS3-style classes then maybe we would want to continue having tables in the future.  But also basically anything on npm could be a data source... The only thing that really makes it AS-specific is taking our cache class as an option...

File: docs/source/config.json
Comment (glasser) [inline docs/source/config.json:20]: Lowercase D looks more consistent?

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:188]:
```suggestion
  constructor(options: { cache: KeyValueCache<string>; contextValue: string }) {
  //highlight-end
    // the necessary arguments for HTTPDataSource
    const pool = new Pool(baseURL);
    super(baseURL, { pool });

    // We need to call the initialize method in our data source's
    // constructor, passing in our cache and contextValue.
    this.initialize({ cache: options.cache, context: options.contextValue }); //highlight-line
```

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-data.mdx:180]: `contextValue` shouldn't be a string here, should be a `TContext` probably, or switched over to `token` or whatever this datasource actually considers useful (as mentioned in previous comments).

Note a change here should be reflected in my suggestion below.

Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:180]: Updating these all to token, so the full snippet is now:

```ts
import { ApolloServer } from '@apollo/server';
import { startStandaloneServer } from '@apollo/server/standalone';
import { KeyValueCache } from '@apollo/utils.keyvaluecache';
import { Pool } from 'undici';
import { HTTPDataSource } from 'apollo-datasource-http';

class BooksAPI extends HTTPDataSource {
  override baseURL = 'https://movies-api.example.com/';

  //highlight-start
  constructor(options: { cache: KeyValueCache<string>; token: string }) {
    //highlight-end
    // the necessary arguments for HTTPDataSource
    const pool = new Pool(baseURL);
    super(baseURL, { pool });

    // We need to call the initialize method in our data source's
    // constructor, passing in our cache and contextValue.
    //highlight-start
    this.initialize({ cache: options.cache, context: options.token });
    //highlight-end

  async getMovie(id: string): Promise<Movie> {
    return this.get<Movie>(`movies/${encodeURIComponent(id)}`);
  }
}

interface MyContext {
  dataSources: {
    moviesApi: MoviesAPI;
  };
  token?: string;
}

const server = new ApolloServer<MyContext>({
  typeDefs,
  resolvers,
});

const { url } = await startStandaloneServer(server, {
  context: async () => {
    //highlight-start
    const { cache } = server;
    const token = req.headers.token;
    return {
      dataSources: {
        moviesApi: new MoviesAPI({ cache, token }),
      },
      //highlight-end
      token,
    };
  },
});
```

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:47]: I forgot to push after our last feedback round, so sorry about that! For reference, this was updated to:

```ts

export class BookingsDataSource {
  private dbConnection;
  private token;
  private user;

  constructor(options: { token: string }) {
    this.dbConnection = this.initializeDBConnection();
    this.token = options.token;
  }
```

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:56]: This is a great point! I updated the snippet to add `getUser`:

```ts
export class BookingsDataSource {
  private dbConnection;
  private token;
  private user;

  constructor(options: { token: string }) {
    this.dbConnection = this.initializeDBConnection();
    this.token = options.token;
  }

  async initializeDBConnection() {
    // set up our database details, instantiate our connection,
    // and return that database connection
    return dbConnection;
  }

  async getUser() {
    if (!this.user) {
      // cache the user, lookup by token
      this.user = await this.dbConnection.User.findByToken(this.token);
    }
    return this.user;
  }

  async getBooking(bookingId) {
    const user = await this.getUser(this.token);
    if (user) {
      return await this.dbConnection.Booking.findByPk(bookingId);
    } else {
      // handle invalid user
    }
  }

  //... more methods for finding and creating bookings
}
```

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:104]: I agree! Updated to:
```
 dataSources: {
        dogsDB: new DogsDataSource({ cache, token: contextValue.token }),
        catsApi: new CatsAPI({ cache }),
      },
```

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx]: I'll add a TODO here to add a link to the `RESTDataSource` page once it covers the two-layer caching model. For now though, how does the following feel to cover this point:
> - If your data source uses an in-memory cache, the `context` function should create a **new instance of your data source class for each operation**. This ensures that your data source doesn't accidentally share values across requests.

**Comment by glasser** [comment]
@smyrick That looks nice! That was actually Plan A for data sources in AS4 but we felt like it still led to too much complexity around how `dataSources` is sometimes part of the context value and sometimes not.

Maybe it would be nice to publish it as some sort of `@apollosolutions` package with a name like LegacyDataSourcesPlugin or something?

**Comment by smyrick** [comment]
We don't want to encourage developers to continue using data sources in this pattern (being invoked every request) if they are able to migrate so we were thinking of just publishing this as a snippet in the docs that developers could manually copy over to their own code as part of the migration guide. This would help them migrate quickly but we would want devs to delete this code eventually by properly migrating each data source

**Comment by glasser** [comment]
Right, invoking per-request is a concept that only makes sense if the actual data source object is stateful... and while RESTDataSource is, we never actually explained that design anywhere (@rkoron007 is *finally* doing that in #7089!) and none of the third-party data sources we linked (other than the one that's a fork of RESTDataSource) does the two-layer caching that RESTDataSource does.  (Which is basically just a way of letting you not have to use both DataLoader and RESTDataSource together.)

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx]: Took another stab at tweaking this now that I have a better understanding of why (and I'll keep the TODO for adding the link to `RESTDataSource` once that PR is merged):
> If your data source is stateful (e.g., uses an in-memory cache), the `context` function should create a **new** instance of your data source class for **each** operation. This ensures that your data source doesn't accidentally share cached results across requests.

**Comment by rkoron007** [comment]
@smyrick thank you for bringing up your snippet! I agree the best way to incorporate this snippet _and_ encourage folks to migrate is to add your suggestion to the migration guide. I'll make a separate PR for that later and tag everyone so we can continue the discussion there :)

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:65]: getUser doesn't take an argument

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:87]:
```suggestion
DataLoader instances are per-request, so if you use a DataLoader in your data source, ensure you **create a new instance of that class with every request**:
```

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:100]: Realistically speaking you probably do *not* want to set up a DB connection once per request.

Probably you want this class to receive the DB connection in its constructor, and set it up at the top level of your program or something, and show it being passed in with your context function.

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:110]: not really sure what "progress" is in this example

**Review by glasser — CHANGES_REQUESTED** [review]
otherwise looking pretty good!

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:65]: Oh, great catch, thank you!

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:219]: Resolving this with the Modern/Legacy sections

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx:110]: I updated this to be more "product"-centric. How does this feel, instead?

```ts
class ProductsDataSource {
  private dbConnection;

  private batchProducts = new DataLoader(async (ids) => {
    const productList = await db.fetchAllKeys(ids);
    return ids.map((id) => productList.find((product) => product.id === id));
  });

  async getProductFor(id) {
    return this.batchProducts.load(id);
  }
}
```

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:99]: Technically this introduces unnecessary quadratic behavior.  Imagine that ids is 1000 elements long. That means productList generally will be too.  Then each `find` call takes on average 500 comparisons to find the id it is looking for, meaning you end up with 500,000 comparisons!  A better approach would be to put the returned items in some sort of map structure so that you just need to do one lookup per ID at the end.

It's not the end of the world to have this in an example but certainly it's nice to obey best practices. Fortunately in this example, we have the superpower that we can assume our underlying APIs do whatever is most useful :) So we could do something like

```
const productsById = await db.fetchAllKeys(ids);
return ids.map((id) => productsById.get(id));
```

**Review by glasser — APPROVED** [review]
Other than that, this looks great!

File: docs/source/data/fetching-data.mdx
Comment (trevor-scheer) [reply docs/source/data/fetching-data.mdx]: @Rose just a couple fixes for the snippet. We need a constructor to receive and store the connection on the class instance. I also updated the reference to `db` to `this.dbConnection` which is what I think you wanted here.

```ts
import DataLoader from 'dataloader';

class ProductsDataSource {
  private dbConnection;

  constructor(dbConnection) {
    this.dbConnection = dbConnection;
  }

  private batchProducts = new DataLoader(async (ids) => {
    const productList = await this.dbConnection.fetchAllKeys(ids);
    return ids.map((id) => productList.find((product) => product.id === id));
  });

  async getProductFor(id) {
    return this.batchProducts.load(id);
  }
}

// In your server file

// Set up our database, instantiate our connection,
// and return that database connection
const dbConnection = initializeDBConnection();

const { url } = await startStandaloneServer(server, {
  context: async () => {
    return {
      dataSources: {
        // Create a new instance of our data source for every request!
        // (We pass in the database connection because we don't need
        // a new connection for every request.)
        productsDb: new ProductsDataSource(dbConnection), //highlight-line
      },
    };
  },
});
```

**Review by trevor-scheer — APPROVED** [review]

File: docs/source/data/fetching-data.mdx
Comment (rkoron007) [reply docs/source/data/fetching-data.mdx]: Sweet, thank you so much Trevor!

**Pull request merged**
